doc/udev-rules.org

```diff
@@ -0,0 +1,244 @@
+
+"Udev" is an implementation detail, the upper layer should say
+"NameRule"
+
+"Udev name" -> "device name", or "persistent device name"
+
+In LanItems, keep the general interface of item["udev"]["net"]
+so that existing code works, BUT
+  instead of ["net"] which is_a Array<String>
+  make       ["name_rule"] which is_a NameRule
+
+*** copying and identity
+#Items are deeply copied values? NO. GetLanItem and getCurrentItem return the
+original data.
+
+*** LanItems item API
+
+This is strictly speaking above our area of focus, but the current naming is
+so awful that at least a glossary is needed to be able to understand what's
+going on.
+
+**** current -> Integer
+the *index* of the current item
+**** GetLanItem(item_id) => item(item_id) # surprisingly no collisions with "item" lvar
+**** getCurrentItem      => current_item
+**** item_name_rule(item_id) = item(item_id)["udev"]["name_rule"]
+also considering that #item would be an adaptor object
+that would translate #name_rule to ["udev"]["name_rule"]
+Does it need to exist? Nil? NullRule?
+**** current_item_name_rule  =  current_item["udev"]["name_rule"]
+
+*** target API
+This is a sketch of the new API as emerging from the Usage section below
+**** NameRule
+***** @udev [UdevRule]
+***** #matcher= and #matcher(:bus_id or :mac)
+***** #value (case sensitive??)
+***** #name
+udev["NAME"]
+**** NameRules
+***** #pathname
+"/etc/udev/rules.d/70-persistent-net.rules"
+
+*** Usage
+Here I list all the mentions of "udev" in the code
+and sketch out how to write them better.
+**** InstallInfConvertor
+***** AllowUdevModify
+checks if cmdline contains "biosdevname=..."
+**** NetworkLanComplexInclude src/include/network/lan/complex.rb
+***** calls LanItems.update_item_udev_rule!(:bus_id)
+**** Lan#Export
+calls LanItems#Export
+**** lan_auto
+***** ToAY converts the net-udev piece from a hash to an array
+**** NetworkLanHardwareInclude
+it's the Hardware tab
+      device_name = LanItems.current_udev_name
+let's keep that
+**** save_network
+#copy_udev_rules
+
+s390 51* leave that
+
+the rule file needs to be copied from inst-sys to target:
+  need its fs path
+  NameRules#pathname (and use std ruby dirname+basename)
+BTW the https://bugzilla.suse.com/show_bug.cgi?id=293366#c7 comment means
+a mkdir -p is fine
+**** edit_nic_name EditNicName
+is a freshly rewritten class, yay 2013-09 mchf
+well, it is called like EditNicName.new.run
+and its #initialize uses the ugly LanItemsApi
+so does #run
+and #CheckUdevNicName (sic)
+***** to be removed:
+MAC_UDEV_ATTR   = "ATTR{address}".freeze
+BUSID_UDEV_ATTR = "KERNELS".freeze
+***** initialize
+@old_key = current_item_name_rule.matcher
+***** run
+        LanItems.update_item_udev_rule!(udev_type)
+(watch out, uses the ui symbol directly)
+***** CheckUdevNicName
+uses LanItems#GetCurrentName
+  which is GetDeviceName(@current) ... and it never uses the "udev name" which
+  confuses my naming plan :(
+renamed! to check_new_device_name
```

Comment:
I do not like new name and I think here you go to non-relevant part and too deep. At first we should propose that new api. Renaming of non-relevant part is out of scope.